src/odoc/env.ml

```diff
@@ -38,61 +54,83 @@ module Accessible_paths = struct
+     we can check the digest right now. If there's more than one, we defer
+     until further up the call stack *)
+  let check_optional_digest ?digest filename (roots : Model.Root.t list) =
+    match roots, digest with
+    | [root], Some d when Digest.compare d root.digest <> 0 ->
+      let warning = Model.Error.filename_only "Digest mismatch" filename in
+      prerr_endline (Model.Error.to_string warning);
+      roots
+    | _ ->
+      roots
+
+  let find_root t ~filename =
     match Hashtbl.find t.file_map filename with
-    | root -> root
+    | roots -> roots
     | exception Not_found ->
-      let path = find_file_by_name t filename in
-      let root = Root.read path in
-      begin match digest with
-      | Some d when Digest.compare d root.digest <> 0 ->
-        Printf.eprintf
-          "WARNING: digest of %s doesn't match the one expected for file %s\n%!"
-          (Fs.File.to_string path) filename
-      | _ -> ()
-      end;
-      Hashtbl.add t.file_map filename root;
-      Model.Root.Hash_table.add t.root_map root path;
-      root
-
-  let file_of_root t root =
+      let paths = find_file_by_name t filename in
+      (* This could be the empty list *)
+      let roots = List.map Root.read paths in
+      Hashtbl.add t.file_map filename roots;
+      List.iter2 (fun root path ->
+        Model.Root.Hash_table.add t.root_map root path) roots paths;
+      roots
+
+  let rec file_of_root t root =
     try Model.Root.Hash_table.find t.root_map root
     with Not_found ->
-      let r =
+      let _roots =
         match root.file with
         | Page page_name ->
           let filename = "page-" ^ page_name in
-          find_root ~digest:root.digest t ~filename
+          check_optional_digest ~digest:root.digest filename @@ find_root t ~filename
         | Compilation_unit { name; _ } ->
-          find_root ~digest:root.digest t ~filename:name
+          check_optional_digest ~digest:root.digest name @@ find_root t ~filename:name
       in
-      assert (Model.Root.equal root r);
-      Model.Root.Hash_table.find t.root_map r
+      file_of_root t root
 end
 
 let rec lookup_unit ~important_digests ap target_name =
+  let handle_root (root : Model.Root.t) = match root.file with
+    | Compilation_unit {hidden; _} -> Xref.Found {root; hidden}
+    | Page _ -> assert false
+  in
   let find_root ~digest =
-    match Accessible_paths.find_root ap ~filename:target_name ?digest with
-    | exception Not_found -> Not_found
-    | root ->
-      match root.file with
-      | Compilation_unit {hidden; _} -> Xref.Found {root; hidden}
-      | Page _ -> assert false
+    match Accessible_paths.find_root ap ~filename:target_name, digest with
+    | [], _ -> Xref.Not_found
+    | [r], _ -> handle_root r (* Already checked the digest, if one's been specified *)
+    | r :: rs, None ->
+      Printf.fprintf stderr "Warning, ambiguous lookup. Please wrap your libraries. Possible files:\n%!";
```

Comment:
In the future, we will probably want to provide more context about which source file depended on the ambiguous compilation unit. Just a note.

Reply:
Yes, that'd be good. We either need a richer return type or to pass in the location into here. I don't know which is better.